balboa-agent/src/main/scala/com/socrata/balboa/agent/MetricConsumer.scala

```diff
@@ -0,0 +1,280 @@
+          var value: Number = null
+          if (!rawValue.equalsIgnoreCase("null")) {
+            try {
+              if (MetricConsumer.integerPattern.matcher(rawValue).matches) {
+                value = JavaLong.valueOf(rawValue)
+              } else {
+                value = JavaDouble.valueOf(rawValue)
+              }
+            }
+            catch {
+              case e: NumberFormatException =>
+                logger.error(s"NumberFormatException reading metric from record: ${record.toString}", e)
+                BalboaAgentMetrics.errorInvalidValueCounter.inc()
+                break()
+            }
+          }
+          val entityId = record.get(MetricConsumer.ENTITY_ID)
+          val name = record.get(MetricConsumer.NAME)
+          val timestamp = JavaLong.parseLong(record.get(MetricConsumer.TIMESTAMP))
+          val metricType = Metric.RecordType.valueOf(record.get(MetricConsumer.RECORD_TYPE).toUpperCase)
+          results.add(new MetricsRecord(entityId, name, value, timestamp, metricType))
+        }
+        record = readRecord(stream)
+      }
+    } finally {
+      logger.info("Completed (possibly with errors) file {}", filePath)
+      // Percolate the exception up the call stack.... ugh.
+      stream.close()
+    }
+    results
+  }
+
+  /**
+    * Accepts a stream of bytes and deserialize them into a single Metric.
+    *
+    * <b>
+    * This class works under the assumptions that bytes are serialized under the following format
+    * <ul>
+    * <li>0xff - single byte - Beginning mark of a single metrics entry</li>
+    * <li>Sequence of bytes of undetermined length - A utf-8 byte String encoded to bytes of a timestamp of type long.</li>
+    * <li>0xfe - single byte - end of timestamp sequence</li>
+    * <li>Sequence of bytes of undetermined length - A utf-8 byte String encoded to bytes of the entity id</li>
+    * <li>0xfe - single byte - end of entity id byte sequence</li>
+    * <li>Sequence of bytes of undetermined length - A utf-8 byte String encoded to bytes of the metric name</li>
+    * <li>0xfe - single byte - end of metric name byte sequence</li>
+    * <li>Sequence of bytes of undetermined length - A utf-8 byte String encoded to bytes of the metric value. The metric value is of type Number</li>
+    * <li>0xfe - single byte - end of metric value byte sequence</li>
+    * <li>Sequence of bytes of undetermined length - A utf-8 byte String encoded to bytes of the metric type. See [[com.socrata.balboa.metrics.Metric.RecordType]]</li>
+    * <li>0xfe - single byte - end of metric type byte sequence</li>
+    * </ul>
+    * </b>
+    *
+    * @param stream A stream of bytes that represent a Single Metric.
+    * @return A Map representing a single Metric entry. null if end of file has been reached.
+    * @throws IOException If an incomplete metrics record was identified.
+    */
+  @throws[IOException]
+  private def readRecord(stream: InputStream): util.Map[String, String] = {
```

Comment:
Maybe split into two steps. Get the record as a string by streaming to the ff and then parse the complete record?